### PR TITLE
[JENKINS-41004] Do not report credentials with IDs masked by nearer folders

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-41044](https://issues.jenkins-ci.org/browse/JENKINS-41004)

Relates to https://github.com/jenkinsci/credentials-plugin/pull/77

@reviewbybees

### CODE REVIEW

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java [hunk 1]
```diff
@@ -119,21 +124,28 @@
...
                                                           @NonNull List<DomainRequirement> domainRequirements) {
-        if (authentication == null) {
-            authentication = ACL.SYSTEM;
-        }
         List<C> result = new ArrayList<C>();
+        Set<String> ids = new HashSet<String>();
         if (ACL.SYSTEM.equals(authentication)) {
             while (itemGroup != null) {
                 if (itemGroup instanceof AbstractFolder) {
+                    if (!((AbstractFolder) itemGroup).getACL().hasPermission(authentication, USE_ITEM)) {
```
Comment (rsandell) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java]: Well, you just verified three lines up that `authentication` is equal to `ACL.SYSTEM` so this seems redundant to me.

**Comment by rsandell** [comment]
:bug: missing test that verifies a folder credential is chosen when there is a system credential with the same id.

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java [hunk 2]
```diff
@@ -119,21 +124,25 @@
...
-                        result.addAll(DomainCredentials.getCredentials(
+                        for (C c : DomainCredentials.getCredentials(
                                 property.getDomainCredentialsMap(),
                                 type,
                                 domainRequirements,
-                                CredentialsMatchers.always()));
+                                CredentialsMatchers.always())) {
+                            if (!(c instanceof IdCredentials) || ids.add(((IdCredentials) c).getId())) {
+                                // if IdCredentials, only add if we havent added already
+                                // if not IdCredentials, always add
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java]: Are there any non-`IdCredentials`, really? Should we just deprecate that possibility?

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java [hunk 3]
```diff
@@ -164,6 +173,61 @@
...
+                                domainRequirements,
+                                matcher)) {
+                            if (ids.add(c.getId())) {
+                                result.add(CredentialsNameProvider.name(c), c.getId());
+                            }
+                        }
+                    }
+                }
+                if (itemGroup instanceof Item) {
+                    itemGroup = Item.class.cast(itemGroup).getParent();
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java]:
```java
itemGroup = ((Item) itemGroup).getParent();
```

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java (hunk 2 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java]: I've see one implementation - cannot recall where - so I am not going to pull the plug on them until I have the implementation guidelines written and published

**Comment by stephenc** [comment]
@rsandell we should be good now! (assuming the CI server agrees)

**Pull request merged**